File: push/src/main/java/com/xiaomi/xmsf/utils/ConfigCenter.java
package com.xiaomi.xmsf.utils;

import android.content.ComponentName;
import android.content.Context;
import android.content.Intent;
import android.content.SharedPreferences;
import android.net.Uri;

import com.xiaomi.xmsf.BuildConfig;
import com.xiaomi.xmsf.push.service.XMPushService;
import com.xiaomi.xmsf.push.utils.Configurations;
import com.xiaomi.xmsf.push.utils.IconConfigurations;

import top.trumeet.common.Constants;
import top.trumeet.common.utils.Utils;


/**
 * Push 配置
 * @author zts
 */
public class ConfigCenter {

    private static class LazyHolder {
        volatile static ConfigCenter INSTANCE = new ConfigCenter();
    }


    public static ConfigCenter getInstance() {
        return LazyHolder.INSTANCE;
    }
    public static void setInstance(ConfigCenter configCenter) {
        LazyHolder.INSTANCE = configCenter;
    }

    public ConfigCenter() {
    }

    //using MODE_MULTI_PROCESS emmm.....
    public static SharedPreferences getSharedPreferences(Context context) {
        return context.getSharedPreferences(BuildConfig.APPLICATION_ID + "_preferences", Context.MODE_MULTI_PROCESS);
    }

    public boolean isNotificationOnRegister(Context ctx) {
        return getSharedPreferences(ctx).getBoolean("NotificationOnRegister", false);
    }

    public boolean isShowConfigurationListOnLoaded(Context ctx) {
        return getSharedPreferences(ctx).getBoolean("ShowConfigurationListOnLoaded", false);
    }

    public int getAccessMode(Context ctx) {
        String mode = getSharedPreferences(ctx).getString("AccessMode", "0");
        return Integer.valueOf(mode);
    }

    public boolean isIceboxSupported(Context ctx) {
        return getSharedPreferences(ctx).getBoolean("IceboxSupported", false);
    }

    public Uri getConfigurationDirectory(Context ctx) {
        String uri = getSharedPreferences(ctx).getString("ConfigurationDirectory", null);
        return uri == null ? null : Uri.parse(uri);
    }

    public boolean setConfigurationDirectory(Context ctx, Uri treeUri) {
        return getSharedPreferences(ctx).edit()
                .putString("ConfigurationDirectory", treeUri.toString())
                .commit();
    }

    public String getXMPPServer(Context ctx) {
        return getSharedPreferences(ctx).getString("XMPP_server", null);
    }

    public boolean setXMPPServer(Context ctx, String host) {
        return getSharedPreferences(ctx).edit()
                .putString("XMPP_server", host)
                .commit();
    }

    public boolean isDebugMode() {
        return getSharedPreferences(Utils.getApplication()).getBoolean("DebugMode", false);
    }

    public boolean isShowAllEvents() {
        return getSharedPreferences(Utils.getApplication()).getBoolean("ShowAllEvents", false);
    }

    public boolean isStartForegroundService() {
        return getSharedPreferences(Utils.getApplication()).getBoolean("StartForegroundService", false);
    }

    public void loadConfigurations(Context context) {
        Configurations.getInstance().init(context,
                ConfigCenter.getInstance().getConfigurationDirectory(context));
        IconConfigurations.getInstance().init(context,
                ConfigCenter.getInstance().getConfigurationDirectory(context));
        Intent intent = new Intent();
        intent.setComponent(new ComponentName(context, XMPushService.class));
        intent.setAction(Constants.CONFIGURATIONS_UPDATE_ACTION);
        context.startService(intent);
    }
}

File: runtime/src/main/java/io/github/magisk317/mipush/runtime/app/ConfigCenter.kt
package io.github.magisk317.mipush.runtime.app

import android.content.ComponentName
import android.content.Context
import android.content.Intent
import android.net.Uri
import com.xiaomi.push.service.XMPushService
import io.github.magisk317.mipush.platform.support.Global
import io.github.magisk317.mipush.runtime.core.data.PreferenceRepository
import io.github.magisk317.mipush.common.configurations.IConfigProvider
import io.github.magisk317.mipush.common.configurations.Configurations
import kotlinx.coroutines.flow.first
import kotlinx.coroutines.launch
import io.github.magisk317.mipush.common.Constants

import javax.inject.Inject
import javax.inject.Singleton

/**
 * Push 配置
 */
@Singleton
class ConfigCenter @Inject constructor(
    private val preferenceRepository: PreferenceRepository
) : IConfigProvider {
    constructor() : this(PreferenceRepository())

    init {
        try {
            io.github.magisk317.mipush.common.utils.Singleton.reset(this)
        } catch (t: Throwable) {
            io.github.aakira.napier.Napier.w("Singleton.reset failed for ConfigCenter", t, tag = "ConfigCenter")
        }
    }
    suspend fun isNotificationOnRegisterAsync(): Boolean =
        preferenceRepository.notificationOnRegister.first()

    override suspend fun isShowConfigurationListOnLoadedAsync(): Boolean =
        preferenceRepository.showConfigurationList.first()

    suspend fun getConfigurationDirectoryAsync(): Uri? {
        val uri = preferenceRepository.configDirectory.first()
        return if (uri.isNullOrBlank()) null else Uri.parse(uri)
    }

    suspend fun setConfigurationDirectoryAsync(treeUri: Uri): Boolean {
        preferenceRepository.setConfigDirectory(treeUri.toString())
        return true
    }

    suspend fun getXMPPServerAsync(): String? = preferenceRepository.xmppServer.first()

    suspend fun setXMPPServerAsync(host: String): Boolean {
        preferenceRepository.setXmppServer(host)
        return true
    }

    suspend fun isDebugModeAsync(): Boolean = preferenceRepository.isDebugMode.first()

    suspend fun isShowAllEventsAsync(): Boolean = preferenceRepository.isShowAllEvents.first()

    suspend fun isStartForegroundServiceAsync(): Boolean = preferenceRepository.isStartForeground.first()

    suspend fun shouldStartPushAsForegroundServiceAsync(): Boolean =
        preferenceRepository.startPushAsForegroundService.first()

    suspend fun getHazeBlurRadiusAsync(): Int = preferenceRepository.hazeBlurRadius.first()

    suspend fun setHazeBlurRadiusAsync(radius: Int): Boolean {
        preferenceRepository.setHazeBlurRadius(radius)
        return true
    }

    suspend fun getHazeTintAlphaAsync(): Float = preferenceRepository.hazeTintAlpha.first()

    suspend fun setHazeTintAlphaAsync(alpha: Float): Boolean {
        preferenceRepository.setHazeTintAlpha(alpha)
        return true
    }

    fun loadConfigurations(context: Context) {
        // Use a cached directory value to avoid blocking the calling thread.
        // The configuration directory rarely changes and is set explicitly by the user.
        kotlinx.coroutines.CoroutineScope(kotlinx.coroutines.Dispatchers.IO).launch {
            val directory = getConfigurationDirectoryAsync()
            Configurations.getInstance().init(context, directory)
            Global.iconConfigurations().init(context, directory)
            val intent = Intent()
            intent.component = ComponentName(context, XMPushService::class.java)
            intent.action = Constants.CONFIGURATIONS_UPDATE_ACTION
            context.startService(intent)
        }
    }
}
